src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2311 @@
+        self.sine, self.cosine = RotaryRelativePositionalBias.get_sinusoid_embeddings(
+            config.max_positions, self.embed_dim
+        )
+        self.alpha = nn.Parameter(torch.Tensor(1, self.embed_dim))
+        self.b_param = nn.Parameter(torch.Tensor(1, self.embed_dim))
+        self.register_buffer("_float_tensor", torch.FloatTensor([0.0]))
+
+    @staticmethod
+    def get_sinusoid_embeddings(max_positions: int, embedding_dim: int):
+        half_dim = embedding_dim // 2
+        emb = math.log(10000) / half_dim
+        emb = torch.exp(torch.arange(half_dim, dtype=torch.float) * -emb)
+        emb = torch.arange(max_positions, dtype=torch.float).unsqueeze(1) * emb.unsqueeze(0)
+        return torch.sin(emb), torch.cos(emb)
+
+    def rotary(self, x):
+        n, d = x.size()
+        x1, x2 = torch.chunk(x, 2, dim=-1)
+        if self.sine is None or n > self.sine.size(0):
+            self.sine, self.cosine = RotaryRelativePositionalBias.get_sinusoid_embeddings(n, d)
+            self.max_positions = n
+        self.sine = self.sine.to(self._float_tensor)
+        self.cosine = self.cosine.to(self._float_tensor)
+
+        sin = self.sine[:n]
+        cos = self.cosine[:n]
+        return torch.cat([x1 * cos - x2 * sin, x2 * cos + x1 * sin], dim=1)
+
+    def forward(self, seq_len):
+        rotary_alpha = self.rotary(self.alpha.expand(seq_len, self.embed_dim))
+        rotary_beta = self.rotary(self.b_param.expand(seq_len, self.embed_dim))
+        bias = torch.einsum("mk,nk->mn", rotary_alpha, rotary_beta)
+        return bias
+
+
+# Normalization modules
+# copied from original Mega repo without modification except variable names
+class ScaleNorm(nn.Module):
+    def __init__(self, dim, eps=1e-6, affine=True):
+        super().__init__()
+        self.dim = dim
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.scalar = nn.Parameter(torch.Tensor(1))
+        else:
+            self.register_parameter("scalar", None)
+
+    def forward(self, input):
+        mean_square = torch.mean(torch.square(input), dim=self.dim, keepdim=True)
+        if self.scalar is not None:
+            input = self.scalar * input
+
+        output = input * torch.rsqrt(mean_square + self.eps)
+        return output
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, number_features, eps=1e-6, affine=True):
+        super().__init__()
+        self.num_features = number_features
+        self.eps = eps
+        self.affine = affine
+        if affine:
+            self.weight = nn.Parameter(torch.Tensor(self.num_features))
+        else:
+            self.register_parameter("weight", None)
+
+    def forward(self, input):
+        mean_square = torch.mean(torch.square(input), dim=-1, keepdim=True)
+        if self.weight is not None:
+            input = input * self.weight
+
+        input * torch.rsqrt(mean_square + self.eps)
+        return input
+
+
+class MegaSequenceNorm(nn.Module):
```

Comment:
Small nit, note really a big fan of this, would rather just index over `NORM2FN` especially because the paper only uses 1. (people could choose the layer norm and we just have to do something along the lines of 
```python 
self.layer_norm = NORM2FN[config.norm]` 
```
like we do for the activation fuctions 😉 see [here](https://github.com/ArthurZucker/transformers/blob/ArthurZucker/issue21300/src/transformers/models/mobilebert/modeling_mobilebert.py#L170)